File: 05-trees-basic/03-tree-traversal/csharp/TreeTraversalDemo.cs
// 03 樹走訪示範（C#）/ Tree traversal demo (C#).  // Bilingual file header.

using System;  // Provide basic runtime types.  
using System.Collections.Generic;  // Provide List<T> for traversal outputs and stack/queue storage.  

namespace TreeTraversalUnit  // Keep this unit isolated within its own namespace.  
{  // Open namespace scope.  
    internal static class TreeTraversalDemo  // Host BinaryTree implementation and traversal helpers.  
    {  // Open class scope.  
        internal sealed class Node  // Binary tree node (value + left child + right child).  
        {  // Open class scope.  
            public Node(int value)  // Initialize node with a value and empty children.  
            {  // Open constructor scope.  
                Value = value;  // Store node value.  
                Left = null;  // Initialize left child as missing.  
                Right = null;  // Initialize right child as missing.  
            }  // Close constructor scope.  

            public int Value { get; }  // Expose node value as read-only.  
            public Node? Left { get; set; }  // Left child pointer (null means no child).  
            public Node? Right { get; set; }  // Right child pointer (null means no child).  
        }  // Close Node.  

        internal readonly struct TraversalSummary  // Summarize recursive/iterative traversal outputs for printing/tests.  
        {  // Open struct scope.  
            public TraversalSummary(IReadOnlyList<int> preorderRecursive, IReadOnlyList<int> preorderIterative, IReadOnlyList<int> inorderRecursive, IReadOnlyList<int> inorderIterative, IReadOnlyList<int> postorderRecursive, IReadOnlyList<int> postorderIterative, IReadOnlyList<int> levelOrder)  // Construct immutable summary.  
            {  // Open constructor scope.  
                PreorderRecursive = preorderRecursive;  // Store recursive preorder.  
                PreorderIterative = preorderIterative;  // Store iterative preorder.  
                InorderRecursive = inorderRecursive;  // Store recursive inorder.  
                InorderIterative = inorderIterative;  // Store iterative inorder.  
                PostorderRecursive = postorderRecursive;  // Store recursive postorder.  
                PostorderIterative = postorderIterative;  // Store iterative postorder.  
                LevelOrder = levelOrder;  // Store level-order traversal.  
            }  // Close constructor scope.  

            public IReadOnlyList<int> PreorderRecursive { get; }  // Preorder via recursion.  
            public IReadOnlyList<int> PreorderIterative { get; }  // Preorder via stack.  
            public IReadOnlyList<int> InorderRecursive { get; }  // Inorder via recursion.  
            public IReadOnlyList<int> InorderIterative { get; }  // Inorder via stack.  
            public IReadOnlyList<int> PostorderRecursive { get; }  // Postorder via recursion.  
            public IReadOnlyList<int> PostorderIterative { get; }  // Postorder via stack+visited.  
            public IReadOnlyList<int> LevelOrder { get; }  // Level-order via BFS queue.  
        }  // Close TraversalSummary.  

        internal sealed class BinaryTree  // A binary tree wrapper that holds a root node.  
        {  // Open class scope.  
            private readonly Node? _root;  // Store root reference (null means empty tree).  

            public BinaryTree(Node? root)  // Initialize tree with an optional root.  
            {  // Open constructor scope.  
                _root = root;  // Store root reference.  
            }  // Close constructor scope.  

            public static BinaryTree FromLevelOrder(int?[] values)  // Build a tree from level-order array representation (null means missing node).  
            {  // Open method scope.  
                if (values is null || values.Length == 0)  // Treat null/empty array as an empty tree.  
                {  // Open empty branch.  
                    return new BinaryTree(root: null);  // Return empty tree.  
                }  // Close empty branch.  

                var nodes = new Node?[values.Length];  // Allocate parallel node array (null for missing).  
                for (int i = 0; i < values.Length; i++)  // First pass: create nodes for non-null values.  
                {  // Open loop scope.  
                    int? v = values[i];  // Read value at index i.  
                    nodes[i] = v.HasValue ? new Node(v.Value) : null;  // Create node or keep null.  
                }  // Close loop scope.  

                for (int i = 0; i < values.Length; i++)  // Second pass: link children using array indices.  
                {  // Open loop scope.  
                    Node? node = nodes[i];  // Read node at index i.  
                    if (node is null)  // Skip missing nodes.  
                    {  // Open skip branch.  
                        continue;  // Move to next index.  
                    }  // Close skip branch.  
                    int leftIndex = 2 * i + 1;  // Compute left child index.  
                    int rightIndex = 2 * i + 2;  // Compute right child index.  
                    if (leftIndex < nodes.Length)  // Link left child when index is in range.  
                    {  // Open left link scope.  
                        node.Left = nodes[leftIndex];  // Assign left child (may be null).  
                    }  // Close left link scope.  
                    if (rightIndex < nodes.Length)  // Link right child when index is in range.  
                    {  // Open right link scope.  
                        node.Right = nodes[rightIndex];  // Assign right child (may be null).  
                    }  // Close right link scope.  
                }  // Close loop scope.  

                return new BinaryTree(root: nodes[0]);  // Root is the node at index 0 (may be null).  
            }  // Close FromLevelOrder.  

            public List<int> PreorderRecursive()  // Return preorder traversal via recursion (root, left, right).  
            {  // Open method scope.  
                var outList = new List<int>();  // Accumulate traversal result.  
                WalkPreorder(_root, outList);  // Walk from root.  
                return outList;  // Return traversal list.  
            }  // Close PreorderRecursive.  

            private static void WalkPreorder(Node? node, List<int> outList)  // Recursive preorder helper.  
            {  // Open method scope.  
                if (node is null)  // Empty subtree contributes nothing.  
                {  // Open base case.  
                    return;  // Stop recursion.  
                }  // Close base case.  
                outList.Add(node.Value);  // Visit root first.  
                WalkPreorder(node.Left, outList);  // Visit left subtree.  
                WalkPreorder(node.Right, outList);  // Visit right subtree.  
            }  // Close WalkPreorder.  

            public List<int> InorderRecursive()  // Return inorder traversal via recursion (left, root, right).  
            {  // Open method scope.  
                var outList = new List<int>();  // Accumulate traversal result.  
                WalkInorder(_root, outList);  // Walk from root.  
                return outList;  // Return traversal list.  
            }  // Close InorderRecursive.  

            private static void WalkInorder(Node? node, List<int> outList)  // Recursive inorder helper.  
            {  // Open method scope.  
                if (node is null)  // Empty subtree contributes nothing.  
                {  // Open base case.  
                    return;  // Stop recursion.  
                }  // Close base case.  
                WalkInorder(node.Left, outList);  // Visit left subtree first.  
                outList.Add(node.Value);  // Visit root.  
                WalkInorder(node.Right, outList);  // Visit right subtree.  
            }  // Close WalkInorder.  

            public List<int> PostorderRecursive()  // Return postorder traversal via recursion (left, right, root).  
            {  // Open method scope.  
                var outList = new List<int>();  // Accumulate traversal result.  
                WalkPostorder(_root, outList);  // Walk from root.  
                return outList;  // Return traversal list.  
            }  // Close PostorderRecursive.  

            private static void WalkPostorder(Node? node, List<int> outList)  // Recursive postorder helper.  
            {  // Open method scope.  
                if (node is null)  // Empty subtree contributes nothing.  
                {  // Open base case.  
                    return;  // Stop recursion.  
                }  // Close base case.  
                WalkPostorder(node.Left, outList);  // Visit left subtree first.  
                WalkPostorder(node.Right, outList);  // Visit right subtree next.  
                outList.Add(node.Value);  // Visit root last.  
            }  // Close WalkPostorder.  

            public List<int> PreorderIterative()  // Return preorder traversal via explicit stack (root, left, right).  
            {  // Open method scope.  
                if (_root is null)  // Empty tree has empty traversal.  
                {  // Open empty case.  
                    return new List<int>();  // Return empty list.  
                }  // Close empty case.  

                var outList = new List<int>();  // Accumulate traversal result.  
                var stack = new List<Node> { _root };  // Use List as stack; start with root.  
                while (stack.Count > 0)  // Process until stack is empty.  
                {  // Open loop scope.  
                    int lastIndex = stack.Count - 1;  // Compute last index for pop.  
                    Node node = stack[lastIndex];  // Read top node.  
                    stack.RemoveAt(lastIndex);  // Pop top node (LIFO).  
                    outList.Add(node.Value);  // Visit node when popped (preorder).  
                    if (node.Right is not null)  // Push right child first so left is processed first.  
                    {  // Open right push scope.  
                        stack.Add(node.Right);  // Push right child.  
                    }  // Close right push scope.  
                    if (node.Left is not null)  // Push left child last.  
                    {  // Open left push scope.  
                        stack.Add(node.Left);  // Push left child.  
                    }  // Close left push scope.  
                }  // Close loop scope.  
                return outList;  // Return traversal list.  
            }  // Close PreorderIterative.  

            public List<int> InorderIterative()  // Return inorder traversal via explicit stack (left, root, right).  
            {  // Open method scope.  
                var outList = new List<int>();  // Accumulate traversal result.  
                var stack = new List<Node>();  // Use List as stack to store the path to current node.  
                Node? node = _root;  // Start from root.  

                while (node is not null || stack.Count > 0)  // Continue while there is a node to visit or stack is non-empty.  
                {  // Open loop scope.  
                    while (node is not null)  // Walk left as far as possible.  
                    {  // Open left-walk scope.  
                        stack.Add(node);  // Push current node so we can come back later.  
                        node = node.Left;  // Move to left child.  
                    }  // Close left-walk scope.  
                    int lastIndex = stack.Count - 1;  // Compute last index for pop.  
                    node = stack[lastIndex];  // Pop the next node whose left subtree is done.  
                    stack.RemoveAt(lastIndex);  // Remove popped element.  
                    outList.Add(node.Value);  // Visit node (inorder).  
                    node = node.Right;  // Move to right subtree next.  
                }  // Close loop scope.  

                return outList;  // Return traversal list.  
            }  // Close InorderIterative.  

            private readonly struct Frame  // Store a node plus a visited flag for iterative postorder.  
            {  // Open struct scope.  
                public Frame(Node node, bool visited)  // Construct immutable frame.  
                {  // Open constructor scope.  
                    Node = node;  // Store node.  
                    Visited = visited;  // Store visited flag.  
                }  // Close constructor scope.  

                public Node Node { get; }  // Node pointer for this frame.  
                public bool Visited { get; }  // True means children were already processed.  
            }  // Close Frame.  

            public List<int> PostorderIterative()  // Return postorder traversal via explicit stack (left, right, root).  
            {  // Open method scope.  
                if (_root is null)  // Empty tree has empty traversal.  
                {  // Open empty case.  
                    return new List<int>();  // Return empty list.  
                }  // Close empty case.  

                var outList = new List<int>();  // Accumulate traversal result.  
                var stack = new List<Frame> { new Frame(_root, visited: false) };  // Store (node, visited) frames.  
                while (stack.Count > 0)  // Process until stack is empty.  
                {  // Open loop scope.  
                    int lastIndex = stack.Count - 1;  // Compute last index for pop.  
                    Frame f = stack[lastIndex];  // Read top frame.  
                    stack.RemoveAt(lastIndex);  // Pop top frame.  
                    if (f.Visited)  // When visited==true, children were already processed.  
                    {  // Open visited branch.  
                        outList.Add(f.Node.Value);  // Visit node last (postorder).  
                        continue;  // Move to next frame.  
                    }  // Close visited branch.  
                    stack.Add(new Frame(f.Node, visited: true));  // Re-push node marked as visited (to output after children).  
                    if (f.Node.Right is not null)  // Push right child first so left is processed first.  
                    {  // Open right push scope.  
                        stack.Add(new Frame(f.Node.Right, visited: false));  // Process right subtree.  
                    }  // Close right push scope.  
                    if (f.Node.Left is not null)  // Push left child last.  
                    {  // Open left push scope.  
                        stack.Add(new Frame(f.Node.Left, visited: false));  // Process left subtree first.  
                    }  // Close left push scope.  
                }  // Close loop scope.  
                return outList;  // Return traversal list.  
            }  // Close PostorderIterative.  

            public List<int> LevelOrder()  // Return level-order traversal (BFS) via queue.  
            {  // Open method scope.  
                if (_root is null)  // Empty tree has empty traversal.  
                {  // Open empty case.  
                    return new List<int>();  // Return empty list.  
                }  // Close empty case.  

                var queue = new List<Node> { _root };  // Use List as queue storage; start with root.  
                int head = 0;  // Maintain head index so dequeue is O(1) without shifting.  
                var outList = new List<int>();  // Accumulate traversal result.  

                while (head < queue.Count)  // Process queue until head reaches the end.  
                {  // Open loop scope.  
                    Node node = queue[head];  // Read next node to process.  
                    head += 1;  // Advance head (logical dequeue).  
                    outList.Add(node.Value);  // Visit node in BFS order.  
                    if (node.Left is not null)  // Enqueue left child when present.  
                    {  // Open left enqueue scope.  
                        queue.Add(node.Left);  // Append to queue tail.  
                    }  // Close left enqueue scope.  
                    if (node.Right is not null)  // Enqueue right child when present.  
                    {  // Open right enqueue scope.  
                        queue.Add(node.Right);  // Append to queue tail.  
                    }  // Close right enqueue scope.  
                }  // Close loop scope.  

                return outList;  // Return traversal list.  
            }  // Close LevelOrder.  

            public TraversalSummary Summarize()  // Build a snapshot summary for demos/tests.  
            {  // Open method scope.  
                return new TraversalSummary(  // Construct immutable summary record.  
                    preorderRecursive: PreorderRecursive(),  // Compute recursive preorder.  
                    preorderIterative: PreorderIterative(),  // Compute iterative preorder.  
                    inorderRecursive: InorderRecursive(),  // Compute recursive inorder.  
                    inorderIterative: InorderIterative(),  // Compute iterative inorder.  
                    postorderRecursive: PostorderRecursive(),  // Compute recursive postorder.  
                    postorderIterative: PostorderIterative(),  // Compute iterative postorder.  
                    levelOrder: LevelOrder()  // Compute level-order BFS.  
                );  // Close summary construction.  
            }  // Close Summarize.  
        }  // Close BinaryTree.  
    }  // Close class scope.  
}  // Close namespace scope.  

File: 05-trees-basic/03-tree-traversal/java/TreeTraversalDemo.java
// 03 樹走訪示範（Java）/ Tree traversal demo (Java).  // Bilingual file header.

import java.util.ArrayList;  // Use ArrayList to build traversal outputs and to implement stacks/queues.  
import java.util.Arrays;  // Use Arrays.toString for readable demo output.  
import java.util.List;  // Use List for type clarity when collecting values.  

public final class TreeTraversalDemo {  // Provide traversal implementations (recursive vs iterative) for study.  
    private TreeTraversalDemo() {  // Prevent instantiation because this class is a static container.  
        // No instances.  // Clarify intent.  
    }  // Close constructor.  

    public static final class Node {  // Binary tree node (value + left child + right child).  
        public final int value;  // Store node value.  
        public Node left;  // Store left child pointer (null means no child).  
        public Node right;  // Store right child pointer (null means no child).  

        public Node(int value) {  // Initialize node with a value and empty children.  
            this.value = value;  // Assign value.  
            this.left = null;  // Initialize left child as missing.  
            this.right = null;  // Initialize right child as missing.  
        }  // Close constructor.  
    }  // End Node.  

    public static final class TraversalSummary {  // Summarize recursive/iterative traversal outputs for printing/tests.  
        public final int[] preorderRecursive;  // Preorder traversal via recursion.  
        public final int[] preorderIterative;  // Preorder traversal via explicit stack.  
        public final int[] inorderRecursive;  // Inorder traversal via recursion.  
        public final int[] inorderIterative;  // Inorder traversal via explicit stack.  
        public final int[] postorderRecursive;  // Postorder traversal via recursion.  
        public final int[] postorderIterative;  // Postorder traversal via explicit stack.  
        public final int[] levelOrder;  // Level-order traversal via BFS queue.  

        public TraversalSummary(int[] preorderRecursive, int[] preorderIterative, int[] inorderRecursive, int[] inorderIterative, int[] postorderRecursive, int[] postorderIterative, int[] levelOrder) {  // Construct immutable summary.  
            this.preorderRecursive = preorderRecursive;  // Store recursive preorder.  
            this.preorderIterative = preorderIterative;  // Store iterative preorder.  
            this.inorderRecursive = inorderRecursive;  // Store recursive inorder.  
            this.inorderIterative = inorderIterative;  // Store iterative inorder.  
            this.postorderRecursive = postorderRecursive;  // Store recursive postorder.  
            this.postorderIterative = postorderIterative;  // Store iterative postorder.  
            this.levelOrder = levelOrder;  // Store level-order traversal.  
        }  // Close constructor.  
    }  // End TraversalSummary.  

    public static final class BinaryTree {  // A binary tree wrapper that holds a root node.  
        private final Node root;  // Store root reference (null means empty tree).  

        public BinaryTree(Node root) {  // Initialize tree with an optional root.  
            this.root = root;  // Store root reference.  
        }  // Close constructor.  

        public static BinaryTree fromLevelOrder(Integer[] values) {  // Build a tree from level-order array representation (null means missing node).  
            if (values == null || values.length == 0) {  // Treat null/empty array as an empty tree.  
                return new BinaryTree(null);  // Return empty tree.  
            }  // Close validation.  

            Node[] nodes = new Node[values.length];  // Allocate parallel node array (null for missing).  
            for (int i = 0; i < values.length; i++) {  // First pass: create nodes for non-null values.  
                Integer v = values[i];  // Read value at index i.  
                nodes[i] = (v == null) ? null : new Node(v);  // Create node or keep null.  
            }  // Close first pass.  

            for (int i = 0; i < values.length; i++) {  // Second pass: link children using array indices.  
                Node node = nodes[i];  // Read node at index i.  
                if (node == null) {  // Skip missing nodes.  
                    continue;  // Move to next index.  
                }  // Close skip branch.  
                int leftIndex = 2 * i + 1;  // Compute left child index.  
                int rightIndex = 2 * i + 2;  // Compute right child index.  
                if (leftIndex < nodes.length) {  // Link left child when index is in range.  
                    node.left = nodes[leftIndex];  // Assign left child (may be null).  
                }  // Close left link.  
                if (rightIndex < nodes.length) {  // Link right child when index is in range.  
                    node.right = nodes[rightIndex];  // Assign right child (may be null).  
                }  // Close right link.  
            }  // Close linking pass.  

            return new BinaryTree(nodes[0]);  // Root is at index 0 (may be null).  
        }  // End fromLevelOrder().  

        public int[] preorderRecursive() {  // Return preorder traversal via recursion (root, left, right).  
            List<Integer> out = new ArrayList<>();  // Accumulate traversal result.  
            preorderWalk(this.root, out);  // Walk from root.  
            return toIntArray(out);  // Convert to primitive array.  
        }  // End preorderRecursive().  

        private static void preorderWalk(Node node, List<Integer> out) {  // Recursive preorder helper.  
            if (node == null) {  // Empty subtree contributes nothing.  
                return;  // Stop recursion.  
            }  // Close base case.  
            out.add(node.value);  // Visit root first.  
            preorderWalk(node.left, out);  // Visit left subtree.  
            preorderWalk(node.right, out);  // Visit right subtree.  
        }  // End preorderWalk().  

        public int[] inorderRecursive() {  // Return inorder traversal via recursion (left, root, right).  
            List<Integer> out = new ArrayList<>();  // Accumulate traversal result.  
            inorderWalk(this.root, out);  // Walk from root.  
            return toIntArray(out);  // Convert to primitive array.  
        }  // End inorderRecursive().  

        private static void inorderWalk(Node node, List<Integer> out) {  // Recursive inorder helper.  
            if (node == null) {  // Empty subtree contributes nothing.  
                return;  // Stop recursion.  
            }  // Close base case.  
            inorderWalk(node.left, out);  // Visit left subtree first.  
            out.add(node.value);  // Visit root.  
            inorderWalk(node.right, out);  // Visit right subtree.  
        }  // End inorderWalk().  

        public int[] postorderRecursive() {  // Return postorder traversal via recursion (left, right, root).  
            List<Integer> out = new ArrayList<>();  // Accumulate traversal result.  
            postorderWalk(this.root, out);  // Walk from root.  
            return toIntArray(out);  // Convert to primitive array.  
        }  // End postorderRecursive().  

        private static void postorderWalk(Node node, List<Integer> out) {  // Recursive postorder helper.  
            if (node == null) {  // Empty subtree contributes nothing.  
                return;  // Stop recursion.  
            }  // Close base case.  
            postorderWalk(node.left, out);  // Visit left subtree first.  
            postorderWalk(node.right, out);  // Visit right subtree next.  
            out.add(node.value);  // Visit root last.  
        }  // End postorderWalk().  

        public int[] preorderIterative() {  // Return preorder traversal via explicit stack (root, left, right).  
            if (this.root == null) {  // Empty tree has empty traversal.  
                return new int[0];  // Return empty array.  
            }  // Close empty case.  

            List<Integer> out = new ArrayList<>();  // Accumulate traversal result.  
            List<Node> stack = new ArrayList<>();  // Use ArrayList as a stack.  
            stack.add(this.root);  // Push root onto stack.  

            while (!stack.isEmpty()) {  // Process until stack is empty.  
                Node node = stack.remove(stack.size() - 1);  // Pop top node (LIFO).  
                out.add(node.value);  // Visit node when popped (preorder).  
                if (node.right != null) {  // Push right child first so left is processed first.  
                    stack.add(node.right);  // Push right child.  
                }  // Close right push.  
                if (node.left != null) {  // Push left child last.  
                    stack.add(node.left);  // Push left child.  
                }  // Close left push.  
            }  // Close loop.  

            return toIntArray(out);  // Convert to primitive array.  
        }  // End preorderIterative().  

        public int[] inorderIterative() {  // Return inorder traversal via explicit stack (left, root, right).  
            List<Integer> out = new ArrayList<>();  // Accumulate traversal result.  
            List<Node> stack = new ArrayList<>();  // Use ArrayList as a stack.  
            Node node = this.root;  // Start from root.  

            while (node != null || !stack.isEmpty()) {  // Continue while there is a node to visit or stack is non-empty.  
                while (node != null) {  // Walk left as far as possible.  
                    stack.add(node);  // Push current node so we can come back later.  
                    node = node.left;  // Move to left child.  
                }  // Close left-walk loop.  
                node = stack.remove(stack.size() - 1);  // Pop the next node whose left subtree is done.  
                out.add(node.value);  // Visit node (inorder).  
                node = node.right;  // Move to right subtree next.  
            }  // Close main loop.  

            return toIntArray(out);  // Convert to primitive array.  
        }  // End inorderIterative().  

        private static final class Frame {  // Store a node plus a visited flag for iterative postorder.  
            public final Node node;  // Node pointer for this frame.  
            public final boolean visited;  // True means children were already processed.  

            public Frame(Node node, boolean visited) {  // Construct immutable frame.  
                this.node = node;  // Store node.  
                this.visited = visited;  // Store visited flag.  
            }  // Close constructor.  
        }  // End Frame.  

        public int[] postorderIterative() {  // Return postorder traversal via explicit stack (left, right, root).  
            if (this.root == null) {  // Empty tree has empty traversal.  
                return new int[0];  // Return empty array.  
            }  // Close empty case.  

            List<Integer> out = new ArrayList<>();  // Accumulate traversal result.  
            List<Frame> stack = new ArrayList<>();  // Use ArrayList as a stack of frames.  
            stack.add(new Frame(this.root, false));  // Push root frame (not visited).  

            while (!stack.isEmpty()) {  // Process until stack is empty.  
                Frame f = stack.remove(stack.size() - 1);  // Pop top frame.  
                if (f.visited) {  // When visited==true, children were already processed.  
                    out.add(f.node.value);  // Visit node last (postorder).  
                    continue;  // Move to next frame.  
                }  // Close visited branch.  
                stack.add(new Frame(f.node, true));  // Re-push node marked as visited (to output after children).  
                if (f.node.right != null) {  // Push right child first so left is processed first.  
                    stack.add(new Frame(f.node.right, false));  // Process right subtree.  
                }  // Close right push.  
                if (f.node.left != null) {  // Push left child last.  
                    stack.add(new Frame(f.node.left, false));  // Process left subtree first.  
                }  // Close left push.  
            }  // Close loop.  

            return toIntArray(out);  // Convert to primitive array.  
        }  // End postorderIterative().  

        public int[] levelOrder() {  // Return level-order traversal (BFS) as an int array.  
            if (this.root == null) {  // Empty tree has empty traversal.  
                return new int[0];  // Return empty array.  
            }  // Close empty case.  

            List<Node> queue = new ArrayList<>();  // Use ArrayList as a queue storage.  
            int head = 0;  // Maintain a head index so dequeue is O(1) without shifting.  
            List<Integer> out = new ArrayList<>();  // Accumulate visited values.  

            queue.add(this.root);  // Enqueue root.  
            while (head < queue.size()) {  // Process queue until head reaches the end.  
                Node node = queue.get(head);  // Read next node to process.  
                head += 1;  // Advance head (logical dequeue).  
                out.add(node.value);  // Visit node in BFS order.  
                if (node.left != null) {  // Enqueue left child when present.  
                    queue.add(node.left);  // Append to queue tail.  
                }  // Close left enqueue.  
                if (node.right != null) {  // Enqueue right child when present.  
                    queue.add(node.right);  // Append to queue tail.  
                }  // Close right enqueue.  
            }  // Close BFS loop.  

            return toIntArray(out);  // Convert to primitive array.  
        }  // End levelOrder().  

        public TraversalSummary summarize() {  // Build a summary snapshot for demos/tests.  
            return new TraversalSummary(  // Construct immutable summary record.  
                preorderRecursive(),  // Compute recursive preorder.  
                preorderIterative(),  // Compute iterative preorder.  
                inorderRecursive(),  // Compute recursive inorder.  
                inorderIterative(),  // Compute iterative inorder.  
                postorderRecursive(),  // Compute recursive postorder.  
                postorderIterative(),  // Compute iterative postorder.  
                levelOrder()  // Compute BFS traversal.  
            );  // Close summary construction.  
        }  // End summarize().  

        private static int[] toIntArray(List<Integer> values) {  // Convert List<Integer> to int[] for compact results.  
            int[] out = new int[values.size()];  // Allocate exact-sized primitive array.  
            for (int i = 0; i < values.size(); i++) {  // Copy each element.  
                out[i] = values.get(i);  // Unbox Integer into int.  
            }  // Close loop.  
            return out;  // Return primitive array.  
        }  // End toIntArray().  
    }  // End BinaryTree.  

    public static String formatDemo() {  // Format a small demo output using a sample tree.  
        Integer[] values = new Integer[] {1, 2, 3, 4, 5};  // Sample tree used across this repo for traversal examples.  
        BinaryTree tree = BinaryTree.fromLevelOrder(values);  // Build tree from level-order list.  
        TraversalSummary s = tree.summarize();  // Summarize all traversal variants.  
        StringBuilder sb = new StringBuilder();  // Accumulate output lines efficiently.  
        sb.append("=== Tree Traversal Demo (Java) ===\n");  // Print header.  
        sb.append("level_order input: ").append(Arrays.toString(values)).append("\n");  // Print input array.  
        sb.append("preorder_recursive=").append(Arrays.toString(s.preorderRecursive)).append("\n");  // Print recursive preorder.  
        sb.append("preorder_iterative=").append(Arrays.toString(s.preorderIterative)).append("\n");  // Print iterative preorder.  
        sb.append("inorder_recursive=").append(Arrays.toString(s.inorderRecursive)).append("\n");  // Print recursive inorder.  
        sb.append("inorder_iterative=").append(Arrays.toString(s.inorderIterative)).append("\n");  // Print iterative inorder.  
        sb.append("postorder_recursive=").append(Arrays.toString(s.postorderRecursive)).append("\n");  // Print recursive postorder.  
        sb.append("postorder_iterative=").append(Arrays.toString(s.postorderIterative)).append("\n");  // Print iterative postorder.  
        sb.append("level_order=").append(Arrays.toString(s.levelOrder)).append("\n");  // Print level-order BFS.  
        return sb.toString();  // Return formatted output.  
    }  // End formatDemo().  

    public static void main(String[] args) {  // CLI entry point for the demo.  
        System.out.print(formatDemo());  // Print demo output.  
    }  // End main().  
}  // End TreeTraversalDemo.  

